asf_levies_model/levies.py

```diff
@@ -0,0 +1,1288 @@
+        """Rebalance levy based on inputs.
+
+            Rebalancing is revenue-based, reapportioning revenue using the provided parameters and deriving
+        rates from the provided supply values or customer numbers.
+
+            It is expected that:
+                new_electricity_weight + new_gas_weight + new_tax_weight = 1
+                new_variable_weight_elec + new_fixed_weight_elec = 1 if new_electricity_weight > 0
+                new_variable_weight_gas + new_fixed_weight_gas = 1 if new_gas_weight > 0
+
+            Args
+                new_electricity_weight: float [0, 1] new proportion of levy revenue to be charged to electricity.
+                new_gas_weight: float [0, 1] new proportion of levy revenue to be charged to gas.
+                new_tax_weight: float [0, 1] new proportion of levy revenue to be charged to general taxation.
+                new_variable_weight_elec: float [0, 1] new proportion of levy electricity revenue to charge based on consumption.
+                new_fixed_weight_elec: float [0, 1] new proportion of levy electricity revenue to charge based on customer numbers.
+                new_variable_weight_gas: float [0, 1] new proportion of levy gas revenue to charge based on consumption.
+                new_fixed_weight_gas: float [0, 1] new proportion of levy gas revenue to charge based on customer numbers.
+                supply_gas: float [0, inf) annual gas supply value (MWh).
+                supply_elec: float [0, inf) annual electricity supply value (MWh)
+                customers_gas: int [0, inf) annual gas customers (customer or meter count).
+                customers_elec: int [0, inf) annual electricity customers (customer or meter count).
+                inplace: bool (default: False): whether to update levy instance inplace or return new levy instance.
+
+            Raises:
+                ValueError: if rebalancing fails to maintain total revenue.
+        """
+
+        # Revenue contributions
+        revenue_gas = self.revenue * new_gas_weight
+        revenue_elec = self.revenue * new_electricity_weight
+        revenue_tax = self.revenue * new_tax_weight
+
+        # New variable levy rate
+        new_levy_var_gas = (revenue_gas / supply_gas) * new_variable_weight_gas
+        new_levy_var_elec = (revenue_elec / supply_elec) * new_variable_weight_elec
+
+        # New fixed levy rate
+        new_levy_fixed_gas = (revenue_gas / customers_gas) * new_fixed_weight_gas
+        new_levy_fixed_elec = (revenue_elec / customers_elec) * new_fixed_weight_elec
+
+        if not self._is_revenue_maintained(
+            new_levy_var_gas,
+            new_levy_var_elec,
+            new_levy_fixed_gas,
+            new_levy_fixed_elec,
+            supply_gas,
+            supply_elec,
+            customers_gas,
+            customers_elec,
+            revenue_tax,
+            self.revenue,
+        ):
+            raise ValueError("Rebalancing failed to maintain revenue.")
```

Comment:
```suggestion
            raise ValueError("Rebalancing failed to maintain revenue. (Try: Check that new electricity-gas-tax and fixed-variable weights provided add up to 1, respectively.)")
```
I think that the most likely reason (or only reason?) for revenue not being maintained will be due to new weighting values not adding up to 1, so thought it might be useful to explicitly prompt the user to check this.